File: ContainsDuplicate.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jan  5 17:02:31 2018

@author: xuwh
"""
class Solution:
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        dic = {}
        for i,v in enumerate(nums):
            if v in dic and i - dic[v] <= k:  #检查字典中是否含有键值为v的项
                return True
            dic[v] = i
        return False

res = Solution()
data = [1,2,4,5,6,8,9,13,4,2]
print(res.containsNearbyDuplicate(data,10)) 

File: ContainsDuplicate.cpp
#include<iostream>
#include<stdlib.h>
#include<vector>
#include<unordered_set>
using namespace std;

class Solution{
public:
	bool containsNearbyDuplicate(vector<int>& nums, int k) {
	unordered_set<int> s;
	if (k <= 0) return false;
	if (k >= nums.size()) k = nums.size()-1;
	for (int i = 0;i < nums.size();i++)
	{
		if (i > k) s.erase(nums[i-k-1]);//相当于以K值大小的窗口为限制进行数据的寻找，当超出的时候不断进行数据的消减。
		if (s.find(nums[i]) != s.end()) return true;
		s.insert(nums[i]);
	}
	return false;
    }
}; 
/*
int main(){

	Solution* res = new Solution();
	vector <int> data;
	data.push_back(1);
	data.push_back(2);
	data.push_back(3);
	data.push_back(2);
	bool ret = res->containsNearbyDuplicate(data,3);
	return 0;
}
*/